pkg/pool/pool.go

```diff
@@ -401,3 +502,212 @@ func (p *pool) Introspect() IntrospectResponse {
 	}
 }
+
+// IsManagedResource checks if the IPv4 address is managed by the handler
+func (p *pool) IsManagedResource(resourceID string) bool {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+	for _, resource := range p.usedResources {
+		if resource.ResourceID == resourceID {
+			return true
+		}
+	}
+	return false
+}
+
+// findMinGroup finds the resource group with the lowest number of resources, and returns the group id along with its count of resources
+func findMinGroup(resourceGroups map[string][]Resource) (group string, minCount int) {
+	for groupID, resources := range resourceGroups {
+		if resources == nil || len(resources) == 0 {
+			continue
+		}
+
+		// initialize return values as the first key-value pair; when lower count is found, record the group and minCount
+		if group == "" || len(resources) < minCount {
+			minCount = len(resources)
+			group = groupID
+		}
+	}
+
+	return
+}
+
+// assignResourceFromMinGroup assigns a resource from the resource group that has the lowest number of resources
+func (p *pool) assignResourceFromMinGroup() Resource {
+	groupID, resourceCount := findMinGroup(p.warmResources)
+	if groupID == "" || resourceCount == 0 {
+		return Resource{}
+	}
+
+	resource := p.warmResources[groupID][0]
+	if resourceCount == 1 {
+		p.warmResources[groupID] = nil
+	} else {
+		p.warmResources[groupID] = p.warmResources[groupID][1:]
+	}
+
+	return resource
+}
+
+// assignResourceFromAnyGroup assigns a resource from any resource group with at least one resource available
+func (p *pool) assignResourceFromAnyGroup() Resource {
+	resource := Resource{}
+	for groupID, resources := range p.warmResources {
+		if resources == nil || len(resources) <= 0 {
+			continue
+		}
+
+		resource = resources[0]
+		if len(resources) == 1 {
+			p.warmResources[groupID] = nil
+		} else {
+			p.warmResources[groupID] = resources[1:]
+		}
+		break
+	}
+
+	return resource
+}
+
+// getPDDeviation returns the deviation in number of IPv4 addresses when PD is enabled, taking into account of all PD configuration params
+func (p *pool) getPDDeviation() int {
+	deviationPrefix := 0
+
+	var isWarmIPTargetDefined, isMinIPTargetDefined, isWarmPrefixTargetDefined bool
+	if p.warmPoolConfig.WarmIPTarget > 0 {
+		isWarmIPTargetDefined = true
+	}
+	if p.warmPoolConfig.MinIPTarget > 0 {
+		isMinIPTargetDefined = true
+	}
+	if p.warmPoolConfig.WarmPrefixTarget > 0 {
+		isWarmPrefixTargetDefined = true
+	}
+
+	numExistingWarmResources := numResourcesFromMap(p.warmResources)
+
+	// if neither WarmIPTarget nor MinIPTarget defined but WarmPrefixTarget is defined, calculate deviation in number of prefixes first
+	if !isWarmIPTargetDefined && !isMinIPTargetDefined && isWarmPrefixTargetDefined {
+		// a prefix is free if it has all possible IPs available for allocation, e.g. for /28 prefix it means 16 IPs
+		numFreePrefix, _ := p.findFreeGroup()
+		deviationPrefix = p.warmPoolConfig.WarmPrefixTarget - numFreePrefix - p.pendingCreate/NumIPv4AddrPerPrefix
```

Comment:
Is this `p.pendingCreate/NumIPv4AddrPerPrefix` accurate? Do you need to add 1?

Reply:
That's a good point. Inside the function `getPDDeviation()`, `p.pendingCreate` is always a multiplier of 16 for PD pool. But I will change it to ceiling division just in case.